### PR TITLE
shortcircuits for threads

### LINKED ISSUE #7035: [Bug]: upstream retries continue after the client has disconnected, and go past `max_retries` — one abandoned request produced 10 upstream attempts over ~20 minutes
### What happens

After the client gives up on a request (issue #7034 — a silent upstream), Bifrost keeps retrying that upstream on its own, roughly every 120 s (each attempt lasts until the upstream closes it), well past `max_retries: 2`. In one run this went on for ten attempts of ~120 s each — about twenty minutes — while I was testing another gateway against the same fake provider; it stopped only when the container was stopped.

### How to reproduce

Same setup as #7034. Run scenario S06 (`fb-s06-no-response`), let the client abandon the call at 130 s, then keep the fake provider running and watch its log (`curl http://127.0.0.1:8401/control/log`).

### What I observed (two runs, both 8 Sep 2026 UTC)

- Run at 20:30 UTC: client gone at 130 s. New attempts for `fb-s06-no-response` reached the provider at about +240 s, +360 s and then roughly every 120 s — ten in total, against `max_retries: 2` — until `docker stop`. One of them arrived while a different gateway was mid-test, six minutes after the request that started it.
- Re-run at 21:12 UTC: attempts 3 and 4 logged at +240.0 s and +360.0 s after the client had left; attempt 4's connection dropped about 20 s after it opened — the moment the container was restarted (21:19 UTC).

### What I expected

A request whose client has disconnected is cancelled upstream, and the retry budget holds regardless of who is still listening. Otherwise a single stuck upstream plus a few impatient clients turns into a slow retry storm.

### Environment

As in #7034: `maximhq/bifrost:v2.1.0`, `max_retries: 2`, backoff 500 ms → 5 s.

---

How I found this: [Failover Bench](https://github.com/failoverbench/failoverbench) is an open harness — a fake provider that misbehaves in sixteen documented ways, run against gateways and SDKs with the same baseline settings (2 retries, 30 s timeout, one fallback). Results are published at https://failoverbench.github.io/failoverbench/ and updated on the first Monday of each month (next: 5 Oct 2026). Happy to re-run against a fix or a branch before then, and to adjust the harness if any of the above is a harness mistake. Not affiliated with any vendor.

### PR DESCRIPTION
**Pull request by akshaydeo**
## Summary

Claude Code's server-side conversation threads are bound to the upstream account that creates them. Bifrost's per-request key selection, retries, and fallbacks cannot guarantee a continuation lands on the same account, causing `thread_not_found` errors roughly half the time on multi-key configurations. This PR makes Bifrost explicitly stateless with respect to Anthropic threads: `thread: {"type": "continue"}` requests are refused before the provider call, prompting the client to resend the full conversation and stop sending the thread field for the rest of the session. Create requests pass through but have the `thread` field stripped from the raw body before it reaches the wire, so no orphaned thread state accumulates upstream.

## Changes

- `anthropicRefuseThreadContinue` short-circuit added to the `/v1/messages` route: any parsed `AnthropicMessageRequest` whose `thread.type` is `"continue"` receives a 400 with `details.error_code: "thread_unsupported_request"` before the Bifrost flow starts. Create requests and token-counting requests are never refused.
- `BuildAnthropicResponsesRequestBody` strips the `thread` field from the raw body on both the standard messages path and the `count_tokens` path, so no thread field ever reaches the upstream API.
- `AnthropicMessageErrorStruct` gains an optional `Details *AnthropicMessageErrorDetails` field (`omitempty`) carrying `error_code`, allowing error envelopes to surface machine-readable codes clients use for recovery without breaking existing error shapes.
- `AnthropicMessageErrorDetails` type introduced with an `ErrorCode` string field.
- Tests added for the short-circuit edge cases (continue refused, create falls through, no thread falls through, malformed thread falls through, count_tokens path never refused, wrong request type falls through), for the raw-body thread strip in both normal and count-tokens modes, for the details marshal/omit behavior, and for the full handler integration (streaming and non-streaming refusals, create proceeds).

## Type of change

- [x] Bug fix
- [ ] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Chore/CI

## Affected areas

- [x] Core (Go)
- [x] Transports (HTTP)
- [x] Providers/Integrations
- [ ] Plugins
- [ ] UI (React)
- [ ] Docs

## How to test

```sh
go test ./core/providers/anthropic/...
go test ./transports/bifrost-http/integrations/...
```

To validate end-to-end: send a `POST /v1/messages` with `"thread": {"type": "continue", "previous_message_id": "msg_123"}` and confirm a 400 response with `details.error_code: "thread_unsupported_request"` is returned and the Bifrost provider call is never made. Send the same body with `"thread": {"type": "create"}` and confirm the request proceeds and the `thread` field is absent from the upstream payload.

## Breaking changes

- [ ] Yes
- [x] No

Clients sending `thread: {"type": "continue"}` will now receive an explicit 400 instead of a `thread_not_found` from upstream. This is intentional: the error code `thread_unsupported_request` is the signal for clients (e.g. Claude Code) to replay the full conversation and stop using the thread field. The `AnthropicMessageErrorStruct` change is additive and backward-compatible due to `omitempty`.

## Related issues

Closes #7035

## Security considerations

No auth, secrets, or PII implications. The change prevents thread state from being forwarded to upstream accounts that did not create it, which reduces the risk of cross-account state leakage in multi-key configurations.

## Checklist

- [x] I read `docs/contributing/README.md` and followed the guidelines
- [x] I added/updated tests where appropriate
- [x] I updated documentation where needed
- [x] I verified builds succeed (Go and UI)
- [x] I verified the CI pipeline passes locally if applicable

### CODE REVIEW

**Comment by akshaydeo** [comment]
* **#7274** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7274?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/maximhq/bifrost/7274?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `dev`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: transports/bifrost-http/integrations/anthropic.go [hunk 1]
```diff
@@ -60,6 +61,55 @@ func createAnthropicCompleteRouteConfig(pathPrefix string) RouteConfig {
...
+	anthropicReq, ok := req.(*anthropic.AnthropicMessageRequest)
+	if !ok {
+		return false, nil
+	}
+	// The parser captures unknown top-level fields as compacted RawMessage; a nil
+	// map means the body was never parsed (large-payload mode) and there is
+	// nothing to inspect.
+	threadRaw, ok := anthropicReq.ExtraParams["thread"].(json.RawMessage)
+	if !ok {
+		return false, nil
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/integrations/anthropic.go]: Fixed - LargePayloadMetadata gains a ThreadType field and anthropicRefuseThreadContinue now reads it (via the cached resolveLargePayloadMetadata) when ExtraParams is unavailable in large-payload mode, so large continuations get the same 400 thread_unsupported_request before the provider flow. A parsed thread field always wins over metadata, and an empty ThreadType leaves behavior unchanged. Covered by new edge-case tests (metadata continue refused, create/empty fall through, parsed-thread-wins-over-metadata). Populating ThreadType during metadata extraction is the enterprise-repo follow-up, since the extractor lives there; until it lands, large-payload behavior is unchanged and self-recovering via the upstream thread_not_found replay.

File: transports/bifrost-http/integrations/anthropic.go [hunk 2]
```diff
@@ -60,6 +61,61 @@ func createAnthropicCompleteRouteConfig(pathPrefix string) RouteConfig {
...
+	// on full-replay bodies either way.
+	if strings.HasSuffix(string(ctx.Path()), "/count_tokens") {
+		return false, nil
+	}
+	payload, err := providerUtils.MarshalSorted(&anthropic.AnthropicMessageError{
+		Type: "error",
+		Error: anthropic.AnthropicMessageErrorStruct{
+			Type:    "invalid_request_error",
+			Message: `Bifrost does not keep Anthropic thread state; replay the full conversation with thread: {"type": "create"}.`,
+			Details: &anthropic.AnthropicMessageErrorDetails{ErrorCode: "thread_unsupported_request"},
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/integrations/anthropic.go:107]: This is intentional - it is the standard in-flight state of a cross-module PR in this repo. LargePayloadMetadata.ThreadType and AnthropicMessageErrorDetails ship in this same PR's core module, so no released core version containing them can exist yet, and feature PRs here never hand-bump sibling-module pins. The bump lands through the automated 'transports: update dependencies --skip-ci' commit after core is tagged (see c193745d2 bumping core 1.8.5 -> 1.8.6), and the core -> framework -> plugins -> transports release order guarantees the pin is current before any GOWORK=off release build of transports runs. CI on this PR resolves cross-module references through the Go workspace and is fully green (14/14). Same pattern as PR #7104, which added core/providers/utils/roundtripper.go and consumed it from transports in one PR.

**Comment by akshaydeo** [comment]
### Merge activity

* **Sep 17, 6:00 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/7274).
